### PR TITLE
Extract flood fill algorithm from FillTilemapTool

### PR DESCRIPTION
**Pull request by ChristianGreiner**
https://github.com/AdamsLair/duality/issues/612

### CODE REVIEW

**Comment by Barsonax** [comment]
I myself am not a big fan of static state due to its potential for all kind of nasty side effects (also it keeps it in memory forever unless you null it). It should be possible to modify the code so that static is no longer needed.

Aside from that looks clean.

File: Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs [hunk 1]
```diff
@@ -0,0 +1,213 @@
+﻿using System;
+
+namespace Duality.Plugins.Tilemaps
+{
+	/// <summary>
+	/// A helper class that provides a flood fill algorithm.
+	/// </summary>
+	public static class TilemapFloodFill
```
Comment (Barsonax) [inline Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs]: We are dealing with static state here now due to the static activeFillBuffer field. Remove static from this class and its members and just use a instance of the TilemapFloodFill in FillTilemapTool to solve this.

File: Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	{
+		private static Grid<bool> activeFillBuffer = new Grid<bool>();
```
Comment (Barsonax) [inline Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs]: We are dealing with static state here now due to the static activeFillBuffer field. Remove static from this class and its members and just use a instance of the TilemapFloodFill in FillTilemapTool to solve this.

Comment (ChristianGreiner) [reply Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs]: Yeah, you are right. I tought that this class should be static too.. (see TilemapCulling -> https://github.com/AdamsLair/duality/blob/master/Source/Plugins/Tilemaps/Core/Tilemaps/TilemapCulling.cs).

Comment (ilexp) [reply Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs]: I think I can see the arguments on both sides here. It makes sense to keep it static for consistency reasons, but it also makes sense not to, for the reasons @Barsonax mentioned. I'm personally leaning towards the instance variant, now that I've seen it mentioned. 

One request I'd add is to make sure the same instance of the flood fill algorithm is re-usable for multiple operations, re-using its internal buffers along the way to avoid allocations.

Comment (Barsonax) [reply Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs]: FillTilemapTool is reused for multiple operations right? Then making the buffer a instance field of FillTilemapTool will make it reuse the buffer for multiple operations. If not then it never reused it in the first place before these changes.

Comment (ilexp) [reply Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs]: Yep, the tools are kept alive and re-used. Even if they weren't though, I think an "algorithm class" instance should be re-usable by design whenever possible.

**Comment by Barsonax** [comment]
Excuse me I meant making the buffer a instance field of TilemapFloodFill. My mistake.

**Comment by ChristianGreiner** [comment]
I made some changes (see last commit) @Barsonax @ilexp

**Comment by ChristianGreiner** [comment]
@ilexp Done 👍 Do you agree with the changes?

File: Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+			// Get the base tile for comparison
+			Tile baseTile = tiles[pos.X, pos.Y];
+
+			// Find the topleft-most tile to start with
+			fillAreaTopLeft = FillAreaFindTopLeft(fillBuffer, tiles, pos, baseTile);
+			fillAreaSize = new Point2(1 + pos.X - fillAreaTopLeft.X, 1 + pos.Y - fillAreaTopLeft.Y);
+			pos = fillAreaTopLeft;
+
+			// Run the main part of the algorithm
+			if (maxTileCount <= 0) maxTileCount = int.MaxValue;
```
Comment (BobGneu) [inline Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs]: Given that this we are already leaving off the `{}` on single lined if blocks this should either be moved down and tabbed up or a space should follow so the assignment doesn't seem related.

File: Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+			}
+			while (lastRowLength != 0 && pos.Y < tiles.Height);
+
+			return true;
+		}
+
+		private Point2 FillAreaFindTopLeft(Grid<bool> fillBuffer, Grid<Tile> tiles, Point2 pos, Tile baseTile)
+		{
+			// Find the topleft-most connected matching tile
+			while (true)
```
Comment (BobGneu) [inline Source/Plugins/Tilemaps/Core/Tilemaps/TilemapFloodFill.cs]: given that we have a test for completion shouldn't we rework this to have the test be up here?

**Comment by ilexp** [comment]
@BobGneu Since this PR is primarily about extracting the algorithm implementation, I'd say all improvements to the algorithm itself are entirely optional. Merging now, but further improvements are of course appreciated!
